File: solutions/python/flower-field/flower_field.py
"""Flowerfield."""

ADJACENT = [complex(x, y) for x in (-1, 0, 1) for y in (-1, 0, 1) if x or y]


def annotate(garden: list[str]) -> list[str]:
    """Annotate a garden."""
    height = len(garden)
    width = len(garden[0]) if garden else 0

    if not all(len(row) == width for row in garden):
        raise ValueError("The board is invalid with current input.")

    flowers = set()
    for y, line in enumerate(garden):
        for x, val in enumerate(line):
            if val == "*":
                flowers.add(complex(x, y))
            elif val != " ":
                raise ValueError("The board is invalid with current input.")

    def cell_value(point: complex) -> str:
        """Return the value for one square."""
        if point in flowers:
            return "*"
        count = sum(point + offset in flowers for offset in ADJACENT)
        return str(count) if count else " "

    return [
        "".join(cell_value(complex(x, y)) for x in range(width))
        for y in range(height)
    ]

File: solutions/go/flower-field/flower_field.go
package flowerfield

import "strings"

// Annotate returns an annotated board
func Annotate(board []string) []string {
	flowers := make(map[[2]int]bool)
	for y, row := range board {
		for x, r := range row {
			if r == '*' {
				flowers[[2]int{x, y}] = true
			}
		}
	}

	getCount := func(x, y int) int {
		var c int
		for dx := -1; dx <= 1; dx++ {
			for dy := -1; dy <= 1; dy++ {
				if flowers[[2]int{x + dx, y + dy}] && (dx != 0 || dy != 0) {
					c++
				}
			}
		}
		return c
	}

	out := make([]string, len(board))
	for y := range len(board) {
		var sb strings.Builder
		for x := range len(board[0]) {
			if board[y][x] == '*' {
				sb.WriteRune('*')
			} else if c := getCount(x, y); c > 0 {
				sb.WriteRune(rune('0' + c))
			} else {
				sb.WriteRune(rune(' ' + c))
			}
		}
		out[y] = sb.String()
	}
	return out
}
